### PR TITLE
Switch from R2DBC to Hibernate

### LINKED ISSUE #544: Switch from R2DBC to Hibernate
**Problem**
R2DBC has some stability issues.

**Solution**
We should investigate Hibernate as an alternative. We have a `grpc-hibernate` branch just need to fix the JMeter portion of it and test it thoroughly. Even if we keep R2DBC we should test it out just in case we need to switch.

**Alternatives**
- R2DBC
- Spring Data JDBC

**Additional Context**

### LINKED ISSUE #558: Stalled clients for historical data at higher page size
**Detailed Description**
When many concurrent subscribers retrieve a large amount of historical data, some of them may get stalled and eventually timeout with RESOURCE_EXHAUSTED.

**Actual Behavior**
1. Set page size to 1000
2. Concurrently subscribe to a large topic with 100 subscribers
3. Wait for a few of the subscribers to stall and eventually get RESOURCE_EXHAUSTED

Based upon the logs, it seems the publisher thread (DB) overwhelms either the subscriber thread (GRPC) or the client itself or some buffer in between. See below log when it occurs:

[stalled.log](https://github.com/hashgraph/hedera-mirror-node/files/4256917/stalled.log)

**Expected Behavior**
No clients should stall

**Environment:**
- Java: OpenJDK 11
- OS: Ubuntu 18.04
- Version: v0.6.0-rc1

**Additional Context**
We throttle historical currently to 100TPS so don't experience this issue in production, but we may need to support higher in the future

### PR DESCRIPTION
**Pull request by steven-sheehy**
**Detailed description**:
- Switch from R2DBC to Hibernate
- Adds a cache for entity existence since we can now use `@Cacheable`
- Fix shared poller running when disabled
- Fix shared poller not retrying when database down and never reconnecting
- Fix polling retriever not retrying when database down and immediately failing
- Bump maxPageSize to 1000 since Hibernate can handle it
- Bump retriever timeout to 60s
- Use TopicMessageRetriever for missed messages in case importer is down for a long time and it needs to poll missed
- Disable shared poller during tests except for tests that explicitly need it

**Which issue(s) this PR fixes**:
Fixes #558 
Fixes #544 

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -98,10 +99,10 @@ value, it is recommended to only populate overridden properties in the custom `a
 | `hedera.mirror.grpc.netty.maxInboundMetadataSize`          | 1024                    | The maximum size of metadata allowed to be received                                            |
 | `hedera.mirror.grpc.port`                                  | 5600                    | The GRPC API port                                                                              |
 | `hedera.mirror.grpc.retriever.enabled`                     | true                    | Whether to retrieve historical massages or not                                                 |
-| `hedera.mirror.grpc.retriever.maxPageSize`                 | 200                     | The maximum number of messages the retriever can return in a single call to the database       |
+| `hedera.mirror.grpc.retriever.maxPageSize`                 | 1000                    | The maximum number of messages the retriever can return in a single call to the database       |
```
Comment (Nana-EC) [inline docs/configuration.md]: nit: since we changed from r2dbc you could note this is with a hibernate ORM

File: hedera-mirror-grpc/pom.xml [hunk 2]
```diff
@@ -41,15 +45,6 @@
             <groupId>io.projectreactor.addons</groupId>
             <artifactId>reactor-extra</artifactId>
         </dependency>
-        <dependency>
```
Comment (Nana-EC) [inline hedera-mirror-grpc/pom.xml]: 😭 but ah well

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java [hunk 3]
```diff
@@ -50,4 +63,31 @@
...
+
+    @Override
+    public boolean isNew() {
+        return true;
+    }
+
+    public static class TopicMessageBuilder {
+        private long consensusTimestamp;
+
+        public TopicMessageBuilder consensusTimestamp(Instant consensusTimestamp) {
```
Comment (Nana-EC) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java]: q: did you not need topicNum?
Then maybe additionally sequenceNumber and realm_num?

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java [hunk 4]
```diff
@@ -20,14 +20,18 @@
...
+import com.hedera.mirror.grpc.config.CacheConfiguration;
 import com.hedera.mirror.grpc.domain.Entity;
 
-public interface EntityRepository extends ReactiveCrudRepository<Entity, Long> {
+public interface EntityRepository extends CrudRepository<Entity, Long> {
 
-    @Query("select * from t_entities where entity_shard = $1 and entity_realm = $2 and entity_num = $3 limit 1")
-    Mono<Entity> findByCompositeKey(long shard, long realm, long num);
+    @Cacheable(cacheNames = "entity", cacheManager = CacheConfiguration.ENTITY_CACHE, sync = true)
+    @Query(value = "select * from t_entities where entity_shard = ?1 and entity_realm = ?2 and entity_num = ?3",
```
Comment (Nana-EC) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java]: Just checking that the benefits of the limit 1 are no longer valid or needed here? Just in case you missed it in your refactor

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java [hunk 5]
```diff
@@ -39,37 +40,32 @@
...
-                .metrics()
-                .doOnSubscribe(s -> log.debug("Executing query: {}", filter))
-                .doOnCancel(() -> log.debug("[{}] Cancelled query", filter.getSubscriberId()))
-                .doOnComplete(() -> log.debug("[{}] Completed query", filter.getSubscriberId()))
-                .doOnNext(t -> log.trace("[{}] Next message: {}", filter.getSubscriberId(), t));
+        TypedQuery<TopicMessage> typedQuery = entityManager.createQuery(query);
+        if (filter.hasLimit()) {
+            typedQuery.setMaxResults((int) filter.getLimit());
+        }
+        return typedQuery.getResultList().stream(); // getResultStream()'s cursor doesn't work with reactive streams
```
Comment (Nana-EC) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: q: So do we have other options for getting those metrics? Not sure how crucial the were.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/retriever/PollingTopicMessageRetriever.java [hunk 6]
```diff
@@ -61,36 +61,34 @@ public PollingTopicMessageRetriever(RetrieverProperties retrieverProperties,
...
                 .timeout(retrieverProperties.getTimeout(), scheduler)
                 .doOnCancel(context::onComplete)
                 .doOnComplete(context::onComplete)
-                .doOnNext(context::onNext)
-                .doOnSubscribe(s -> log.info("Starting to poll every {}ms: {}", frequency.toMillis(), filter));
+                .doOnNext(context::onNext);
     }
 
-    private Flux<TopicMessage> poll(PollingContext context) {
+    private Stream<TopicMessage> poll(PollingContext context) {
```
Comment (Nana-EC) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/retriever/PollingTopicMessageRetriever.java]: q: curious, why the change to Stream vs converting to Flux?

File: hedera-mirror-grpc/src/main/resources/application.yml [hunk 7]
```diff
@@ -20,6 +15,8 @@ logging:
   level:
     root: warn
     com.hedera.mirror.grpc: info
+    # org.hibernate.SQL: debug
```
Comment (Nana-EC) [inline hedera-mirror-grpc/src/main/resources/application.yml]: nit: left over from local debugging?

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java (hunk 3 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java]: It's added automatically by Lombok.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java (hunk 4 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java]: True, since it's a native query Hibernate/Spring isn't going to add that automatically. Technically this is a unique composite key so there are never more than one, but will add to be safe.

File: hedera-mirror-grpc/src/main/resources/application.yml (hunk 7 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/resources/application.yml]: Yes and no. I like to keep these particular two around as we use them all the time and it's easier to uncomment than google for their name each time.

File: docs/configuration.md (hunk 1 above)
Comment (steven-sheehy) [reply docs/configuration.md]: I feel specifying the specific ORM in an otherwise generic config option couples it to a particular implementation unnecessarily.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java (hunk 5 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/TopicMessageRepositoryCustomImpl.java]: I found a way by putting it in the retriever.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/retriever/PollingTopicMessageRetriever.java (hunk 6 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/retriever/PollingTopicMessageRetriever.java]: Because the database layer can't be a flux and it wasn't really necessary to have it be reactive. But since I added back the metrics it became useful so converted it back to a flux

File: docs/configuration.md (hunk 1 above)
Comment (Nana-EC) [reply docs/configuration.md]: Understood.

File: hedera-mirror-grpc/src/main/resources/application.yml (hunk 7 above)
Comment (Nana-EC) [reply hedera-mirror-grpc/src/main/resources/application.yml]: Works for me.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java [hunk 8]
```diff
@@ -50,4 +63,31 @@
...
+    }
+
+    @Override
+    public Long getId() {
+        return consensusTimestamp;
+    }
+
+    @Override
+    public boolean isNew() {
+        return true;
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/domain/TopicMessage.java]: `topic_message.consensus_timestamp` can never be null, it's a required field. We don't ever update topic messages, we only insert. Hence why we can avoid the query for existence that JPA does for natural IDs by always returning true for isNew.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java [hunk 9]
```diff
@@ -20,14 +20,18 @@
...
+import com.hedera.mirror.grpc.config.CacheConfiguration;
 import com.hedera.mirror.grpc.domain.Entity;
 
-public interface EntityRepository extends ReactiveCrudRepository<Entity, Long> {
+public interface EntityRepository extends CrudRepository<Entity, Long> {
 
-    @Query("select * from t_entities where entity_shard = $1 and entity_realm = $2 and entity_num = $3 limit 1")
-    Mono<Entity> findByCompositeKey(long shard, long realm, long num);
+    @Cacheable(cacheNames = "entity", cacheManager = CacheConfiguration.ENTITY_CACHE, sync = true)
+    @Query(value = "select * from t_entities where entity_shard = ?1 and entity_realm = ?2 and entity_num = ?3 limit 1",
```
Comment (apeksharma) [inline hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java]: Thinking about new topics.
Say i create a new topic, and immediately subscribe for it. The topic will take ~5-10 seconds to be created in db. So the first lookup will return empty, and also all the subsequent lookups until cache gets invalidated.
I believe the tests may be working since they share the Repo instance, but that won't be the case in prod were importer inserts the data.
If we cache only non-null values, that should work and should be okay (since this isn't anywhere near being the perf bottleneck in our system). Further optimizations can be in future.

Side question for my understanding: does the cache has any internal time based invalidation too if size limit is never reached? Not saying we need it, just random q.

File: hedera-mirror-grpc/src/test/resources/application-default.yml [hunk 10]
```diff
@@ -15,10 +15,13 @@ hedera:
         username: mirror_node
       endTimeInterval: 100ms
       listener:
+        enabled: false
```
Comment (apeksharma) [inline hedera-mirror-grpc/src/test/resources/application-default.yml]: why do we need to disable it?
Let's also add the comment to the config itself.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/service/TopicMessageServiceTest.java [hunk 11]
```diff
@@ -446,13 +456,13 @@ void bothMessagesWithTopicNum() {
     void bothMessagesWithRealmNum() {
         domainBuilder.entity(e -> e.entityRealm(1L)).block();
         domainBuilder.entity(e -> e.entityRealm(2L)).block();
-        domainBuilder.topicMessage(t -> t.realmNum(0)).block();
-        domainBuilder.topicMessage(t -> t.realmNum(1)).block();
+        domainBuilder.topicMessage(t -> t.realmNum(0).sequenceNumber(1)).block();
```
Comment (apeksharma) [inline hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/service/TopicMessageServiceTest.java]: in future,  maybe we can let domainBuilder keep track of seq numbers for topics.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java [hunk 12]
```diff
@@ -103,12 +106,13 @@ void setup() {
...
 
-    private <T> Mono<?> insert(T domainObject) {
-        return databaseClient.insert()
-                .into((Class<T>) domainObject.getClass())
-                .using(domainObject)
-                .fetch()
-                .first()
-                .doOnNext(d -> log.debug("Inserted: {}", domainObject));
+    private Mono<Entity> insert(Entity entity) {
+        return Mono.defer(() -> Mono.just(entityRepository.save(entity)))
```
Comment (apeksharma) [inline hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java]: We should not use grpc server's `Repository` instances (here and in other places) to insert data.
Correct setup, and one analogous to production would be, where writers and readers are isolated. Writers would be analogous to mirror-node-importer, and readers would be grpc server's repository instances (since the tests are to test them).

File: hedera-mirror-grpc/src/test/resources/application-default.yml (hunk 10 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/test/resources/application-default.yml]: Done

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/service/TopicMessageServiceTest.java (hunk 11 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/service/TopicMessageServiceTest.java]: DomainBuilder does keep track of sequenceNumbers for simple case of same sequence number for all topic/realm numbers. But I didn't feel it was worth updating it for separate sequence numbers for multiple topic/realm numbers since it's only done in a couple tests and would make it more complex.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java (hunk 12 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java]: I would disagree with you. Repository should be the only means of interacting with the database for both reading and writing. This is the entire point of a repository: to act as the higher level API for CRUD operations to a backend database. Using a Repository allows us to swap out the implementation easily (e.g. PostgreSQL to BigQuery/MongoDB/etc).

Case in point, if we would've use the repository in the domain builder for R2DBC then we would've minimized our changes when switching to Hibernate.

File: hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java (hunk 9 above)
Comment (steven-sheehy) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java]: Good catch. I've pushed a fix to not cache if null.

There is no cache invalidation for this particular cache. I didn't feel we needed it since entities are never deleted and we aren't using any of the fields that are mutable by the importer.

Comment (apeksharma) [reply hedera-mirror-grpc/src/main/java/com/hedera/mirror/grpc/repository/EntityRepository.java]: This was critical bug we just found, there should be a test for this.

File: hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java (hunk 12 above)
Comment (apeksharma) [reply hedera-mirror-grpc/src/test/java/com/hedera/mirror/grpc/domain/DomainBuilder.java]: The tests right now are violating a major correctness abstraction - there are separate writer (importer) and readers (grpc) in production, but in tests, same instance is both reader and writer.
That coupled with the fact that a feature which majorly interacts with reads/writes - caching - is disabled in tests but enabled in production just seems very wrong to me.
Imo, correctness comes first, then design.
We should strive for correct testing. Whether it's with or without repository, am agnostic to that.
Do or not, up to you, just stating my strong opinion.
